### PR TITLE
[https://nvbugs/6625695][fix] Teacher-force both paths onto one token script via a +BIAS logits processor so…

### PR DESCRIPTION
**Pull request by trtllm-agent**
## Summary
- **Root cause**: Bit-exact greedy token equality across batch compositions is an invalid invariant — batched decode's M=2 GEMM shifts logits 1-2 bf16 ULP and an exact top-2 tie flips argmax into a divergent sentence.
- **Fix**: Teacher-force both paths onto one token script via a +BIAS logits processor so all 16 steps are compared at tolerance 0.4 (floor 0.25, 1% leak 0.42), pair two distinct videos, and remove the waiver.
- **Original test**: `pytest tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py::test_nemotron_nano_v2_vl_video_batch_equivalence -v`
- Automated fix generated by repair-bot

## Test plan
- [x] Verify fix on the same GPU type as the original failure
- [x] Check for regressions in related tests

## Links
- Bug: https://nvbugs/6625695

## Dev Engineer Review

- The test now uses teacher forcing with a shared 16-step token script.
- The `+BIAS` logits processor prevents autoregressive divergence during comparison.
- The `0.4` tolerance accounts for bf16 GEMM differences.
- The test uses two distinct videos and validates token-script adherence and tensor shapes.
- The waiver removal is correctly scoped to the Nemotron Nano V2 VL video batch-equivalence test.
- No public API or configuration changes were introduced.

## QA Engineer Review

- Modified test function: the Nemotron Nano V2 VL video batch-equivalence test.
- The test is covered by the removed entry in `tests/integration/test_lists/waives.txt`.
- Verdict: sufficient.

### CODE REVIEW

**Review by brnguyen2 — COMMENTED** [review]
The rewrite is the right call: bit-exact greedy token equality across batch compositions was never a runtime guarantee, and teacher-forcing both paths onto one script converts a 1-bit argmax coin flip into a continuous, all-16-step logits comparison. Pairing two distinct videos also fixes the fact that the old `world.mp4`-twice setup made the leakage the test claims to detect unobservable. Two things to tighten (inline), plus:

**CI evidence for the waiver removal.** The failure rate this waiver covers is a couple of percent, so one green run doesn't demonstrate much. Please run the stage that actually owns this test (`l0_l40s.yml` → `L40S-PyTorch-2`) explicitly and, ideally, loop the test locally on L40S several dozen times. `/bot run --stage-list "L40S-PyTorch-2"`.

**Not all of the tracked failures are this assertion.** The bug's failure population includes failures inside the shared `nano_llm_model` fixture (executor creation / GPU memory), not the equality assert. This PR does not address those, so the waiver removal may re-expose a residual failure mode under a different signature. Worth saying so in the PR description rather than implying the tracked flake is fully closed.

**Runtime.** The test now runs five `generate` calls (two discovery + two solo + one batched) instead of three, each decoding an 8-frame video. On a stage this file has already hit memory pressure on, that cost is worth a sentence in the description. If you want it back down, the discovery pass could be dropped by scripting a fixed arbitrary token sequence — the comparison doesn't require the script to be the natural greedy trajectory, only that both paths use the same one.

Also: `stream_ptr` is always `None` on the PyTorch backend (`model_engine.py:8293` passes `None`), so the `ExternalStream` branch in the processor is dead code.

File: tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py [hunk 1]
```diff
@@ -492,74 +492,152 @@ def _build_inputs(prompts_subset, media_subset):
...
-        assert b_token_ids == s_token_ids, (
-            f"Request {i}: token_ids differ between batched and separate runs.\n"
-            f"  batched : {b_token_ids}\n"
-            f"  separate: {s_token_ids}"
+        b_logits = b_out.outputs[0].generation_logits.float().cpu()
+        assert b_logits.shape == s_logits.shape, (
+            f"Request {i}: generation_logits shape differs between batched "
+            f"{tuple(b_logits.shape)} and separate {tuple(s_logits.shape)} runs."
+        )
+        assert b_logits.shape[0] == max_tokens, (
```
Comment (brnguyen2) [inline tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py]: This should assert against `len(scripts[i])`, not `max_tokens`. The script comes from a free-running greedy pass capped at `max_tokens`; if that pass stops early on EOS, the script is shorter than 16, both forced runs also stop early, `_assert_follows_script` passes, and this assert then fails with a message about "silently covering fewer steps" that misdescribes what happened. Either compare to `len(scripts[i])` here, or assert `len(script) == max_tokens` right after discovery so the failure names the real cause.

File: tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py [hunk 2]
```diff
@@ -492,74 +492,152 @@ def _build_inputs(prompts_subset, media_subset):
...
+    BIAS = 1.0e4
+
+    def __init__(self, script):
+        self._script = script
+        self._step = 0
+
+    def __call__(self, req_id, logits, token_ids, stream_ptr, client_id):
+        stream = None if stream_ptr is None else torch.cuda.ExternalStream(stream_ptr)
+        with torch.cuda.stream(stream):
+            if self._step < len(self._script):
```
Comment (brnguyen2) [inline tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py]: Indexing the script by a private call counter makes the script alignment depend on the processor being invoked exactly once per decode step. That holds for the current config, but any extra or repeated invocation (a preempted-and-replayed request, an extra context-chunk call) shifts every subsequent step and trips `_assert_follows_script` — a new flake in a test whose whole purpose is de-flaking.

`token_ids` carries the request's tokens so far, so the step can be derived statelessly and is then self-correcting under replay:

```python
def __call__(self, req_id, logits, token_ids, stream_ptr, client_id):
    if self._base_len is None:
        self._base_len = len(token_ids[0])
    step = len(token_ids[0]) - self._base_len
    if step < len(self._script):
        logits[..., self._script[step]] += self.BIAS
```

File: tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py [hunk 3]
```diff
@@ -492,74 +492,152 @@ def _build_inputs(prompts_subset, media_subset):
...
+
+    Adds a large bias at the scripted index instead of masking the rest to
+    `-inf`: argmax becomes the scripted token, yet every other vocab entry
+    survives untouched so callers can still compare raw model output. Because
+    both the batched and the solo run add the *same* bias at the *same* index,
+    the bias cancels in their difference and even the forced index stays
+    comparable.
+    """
+
+    BIAS = 1.0e4
```
Comment (brnguyen2) [inline tests/unittest/_torch/modeling/test_modeling_nemotron_nano_v2_vl.py:511]: The docstring's claim that the bias cancels and "even the forced index stays comparable" depends on the logits being fp32 when processors run (`_torch/modules/logits_processor.py:34` does `logits.float()`). That's true today — fp32 ULP at 1e4 is ~0.001, three orders below the 0.4 tolerance — but in bf16 the ULP at 1e4 is 64, and the forced column would swamp the comparison. Worth noting the fp32 dependency in the docstring, or dropping BIAS to something like 1e3 so the margin is larger.

**Review by yechank-nvidia — APPROVED** [review]
